server/docs/design/block-verification.md

```diff
@@ -0,0 +1,198 @@
+    S->>S: async completeHashing()   
+
+    S->>SV: (5) verifySignature(signature, computedHash, blockNumber)    
+
+    Note over SV,BSM: (6) Compare computed hash and signature
+    alt (7) Verification Fails
+      SV->>BSM: updateBlockStatus(blockNumber, ERROR, SIGNATURE_INVALID      
+    else (8) Verification Succeeds
+      SV->>BSM: updateBlockStatus(blockNumber, VERIFIED, NONE)      
+    end
+    Note over BSM: (9) Follow-up to downstream services
+    end
+
+```
+**Note:** The gray boxes on async hash computation will be implemented later only if we need to improve the latency of the verified blocks stream. (see note at the end of the document)
+## Interfaces
+
+### VerificationHandler
+
+```java
+public interface VerificationHandler {
+  void onBlockItemsReceived(List<BlockItem> blockItems);  
+}
+```
+### BlockHashingSessionFactory
+
+```java
+import java.util.concurrent.ExecutorService;
+
+public interface BlockHashingSessionFactory {
+  BlockHashingSession createSession(List<BlockItem> initialBlockItems, ExecutorService executorService, SignatureVerifier signatureVerifier);
+}
+```
+### BlockHashingSession
+```java
+/* Once hashing is completed internally, calls SignatureVerifier to continue with verification process */
+public interface BlockHashingSession {
+  void addBlockItem(BlockItem item);
+  CompletableFuture<Void> completeHashing(); // triggers final hash computation asynchronously  
+}
+```
+### SignatureVerifier
+```java
+public interface SignatureVerifier {
+  void verifySignature(byte[] signature, byte[] computedHash, long blockNumber);  
+}
+```
+### BlockStatusManager
+```java
+
+public interface BlockStatusManager {
+    void updateBlockStatus(long blockNumber, BlockVerificationStatus status, VerificationError error);
+}
+```
+
+## Enums
+```java
+public enum BlockVerificationStatus {
+  VERIFIED,
+  ERROR
+}
+
+public enum VerificationError {
+  NONE,
+  SIGNATURE_INVALID,
+  SYSTEM_ERROR
+}
+```
+
+## Metrics
+
+- **blocks_received**: Counter of the number of blocks received for verification
+- **blocks_verified**: Counter of the number of blocks verified
+- **blocks_unverified**: Counter of the number of blocks that have not been verified, is the difference between blocks_received and blocks_verified.
+- **blocks_signature_invalid**: Counter of the number of blocks with invalid signatures
+- **blocks_system_error**: Counter of the number of blocks with system errors
+- **block_verification_time**: Histogram of the time taken to verify a block, gives the node operator an idea of the time taken to verify a block.
+- 
+## Exceptions
+
+- **SYSTEM_ERROR:** Issues with node configuration or bugs. The node logs details, updates metrics, and might attempt recovery or halt.
+
+- **SIGNATURE_INVALID:** If verification fails, SIGNATURE_INVALID is used. The block is marked invalid, and the BlockStatusManager triggers error-handling routines (requesting re-sends, removing corrupted data, notifying subscribers, etc.).
```

Comment:
These extra newlines aren't needed and make the raw text a little harder to follow.

Reply:
✅ Done!